src/core/tsi/s2a/record_protocol/s2a_crypter.cc

```diff
@@ -256,3 +248,244 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+ *  authenticating a TLS 1.3 record; the caller must free the buffer using
+ *  gpr_free.
+ *  - sequence: a buffer containing the sequence number of a half connection.
+ *  - sequence_size: the size (in bytes) of the |sequence| buffer.
+ *  - record_header: a buffer containing the header of a TLS 1.3 record.
+ *  - header_size: the size (in bytes) of the |record_header| buffer.
+ *  - payload_size: the size (in bytes) of the payload of a TLS 1.3 record. **/
+static uint8_t* s2a_additional_data(uint8_t* sequence, size_t sequence_size,
+                                    uint8_t* record_header, size_t header_size,
+                                    size_t payload_size) {
+  size_t additional_data_size = sequence_size + header_size;
+  GPR_ASSERT(additional_data_size == kTlsAdditionalDataBytesSize);
+  uint8_t* additional_data =
+      (uint8_t*)gpr_malloc(additional_data_size * sizeof(uint8_t));
+  memcpy(additional_data, sequence, sequence_size);
+  memcpy(additional_data + sequence_size, record_header, header_size);
+  additional_data[11] = payload_size >> 8;
+  additional_data[12] = payload_size & 0xff;
+  return additional_data;
+}
+
+/** This method returns a buffer populated with the nonce used for
+ *  encrypting and decrypting a TLS 1.3 record; the caller must free
+ *  the buffer using gpr_free.
+ *  - crypter: an instance of s2a_crypter.
+ *  - sequence: a buffer containing the sequence number of a half connection.
+ *  - sequence_size: the size (in bytes) of the |sequence| buffer.
+ *  - nonce_size: the number of bytes written to the nonce buffer that is
+ *    returned by the method. **/
+static uint8_t* s2a_nonce(s2a_crypter* crypter, uint8_t* sequence,
+                          size_t sequence_size, size_t* nonce_size) {
+  GPR_ASSERT(crypter != nullptr);
+  size_t max_nonce_size = s2a_max_aead_nonce_size(crypter);
+  GPR_ASSERT(max_nonce_size > sequence_size);
+  GPR_ASSERT(sequence != nullptr);
+  GPR_ASSERT(nonce_size != nullptr);
+  uint8_t* nonce = (uint8_t*)gpr_malloc(max_nonce_size * sizeof(uint8_t));
+  *nonce_size = crypter->out_connection->fixed_nonce_size;
+  memset(nonce, 0, max_nonce_size);
+  memcpy(nonce, crypter->out_connection->fixed_nonce, *nonce_size);
+  for (size_t i = 0; i < sequence_size; i++) {
+    nonce[*nonce_size - sequence_size + i] ^= sequence[i];
+  }
+  return nonce;
+}
+
+grpc_status_code s2a_write_tls13_record(
+    s2a_crypter* crypter, uint8_t record_type, const iovec* unprotected_vec,
+    size_t unprotected_vec_size, iovec protected_record, size_t* bytes_written,
+    char** error_details) {
+  GPR_ASSERT(crypter != nullptr);
+  GPR_ASSERT(crypter->out_connection != nullptr);
+  GPR_ASSERT(crypter->out_connection->initialized);
+  GPR_ASSERT(protected_record.iov_base != nullptr);
+  GPR_ASSERT(bytes_written != nullptr);
+  if ((unprotected_vec == nullptr && unprotected_vec_size > 0) ||
+      (unprotected_vec != nullptr && unprotected_vec_size == 0)) {
```

Comment:
Not sure why you need this second check, having non-null pointer and size == 0 is pretty natural.

Reply:
Removed the second check.